### ISSUE TITLE
Unable to infer initializer's type of generic decodable type

### ISSUE THREAD

**Issue by ishkawa** [ikesyo/Himotoki#22]
In generic decodable type, Swift can't infer the type of initializer. For example:

``` swift
struct PaginatedResponse<T: Decodable where T == T.DecodedType> {
    let id: Int
    let content: T

    static func decode(e: Extractor) -> PaginatedResponse? {
        let create = { PaginatedResponse($0) } // <- Missing argument for parameter 'content' in call

        return build(
            e <| "id",
            e <| "content"
        ).map(create)
    }
}
```

FYI, the code below passes build successfully.

``` swift
struct PaginatedResponse<T: Decodable where T == T.DecodedType> {
    let id: Int
    let content: T

    static func decode(e: Extractor) -> PaginatedResponse? {
        let create = { id, content in
            PaginatedResponse(id: id, content: content)
        }

        return build(
            e <| "id",
            e <| "content"
        ).map(create)
    }
}
```

Do you have any idea to infer the type of initializer in the regular way?

**Comment by ishkawa** [comment]
Ah, this issue is not reproduced in Swift 2 beta 4 :bow: 
I confirmed it by following code:

``` swift
struct GenericResponse<T: Decodable where T == T.DecodedType> {
    let id: Int
    let content: T

    static func decode(e: Extractor) -> GenericResponse? {
        return build(
            e <| "id",
            e <| "content"
        ).map(self.init)
    }
}
```

**Comment by ikesyo** [comment]
Unfortunately, it seems that is a Swift's type inferencer issue related to tuple type with generic types.

As your second sample code of Swift 2 does, passing a static initializer reference directly to `build()` is one possible solution to this, AFAIK. I'm sorry for the inconvenience. :sweat_smile:

**Comment by ikesyo** [comment]
I'm gonna close this since [1.0.0](https://github.com/ikesyo/Himotoki/releases/tag/1.0.0) is released with Swift 2.0 / Xcode 7.0 support.
